035-rebalance-types-scaling-brokers.md

```diff
@@ -0,0 +1,118 @@
+This needs come from how the Cruise Control API endpoints works. There is a specific query parameter which needs that list.
+
+A `KafkaRebalance` custom resource would look like this.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaRebalance
+metadata:
+  name: my-rebalance
+  labels:
+    strimzi.io/cluster: my-cluster
+spec:
+  mode: <value> # full-rebalance, add-broker or remove-broker. Assuming full-rebalance by default if the field is not specified.
+  brokers:
+    - 3
+    - 4
+  goals:
+    - RackAwareGoal
+    - ReplicaCapacityGoal
+```
+
+### Flow
+
+The proposal doesn't assume any automation to run the rebalancing when the cluster is scaled up or down.
+Using the `StatefulSet` for Kafka brokers, the corresponding broker IDs are implied in one direction (added from the higher one up, removed from the higher one down) and the cluster operator could get them for the rebalancing.
+But on the other side, Strimzi is moving to use the [`StrimziPodSet`](https://github.com/strimzi/proposals/pull/44) and towards a ZooKeeper-less cluster support so that the broker IDs are handled differently and the user has the flexibility to scale down by removing specific brokers or scale up by specifying new broker IDs.
+For this reason the proposal is about having the user to specify the added or removed brokers for the rebalancing.
+If needed, it also provides the flexibility to scale up the cluster but running the rebalancing taking into account only a subset of the newly added brokers and leaving others just for new topics.
+The same could happen by running a rebalance for removing brokers to make them "empty" but then not doing the actual scale down.
+
+#### Rebalance on scale up
+
+When scaling up the cluster, the user can go through the following procedure:
+
+1. The user increases the number of replicas through the `spec.kafka.replicas` of the `Kafka` custom resource.
+2. When the scaling up is done, the user creates a `KafkaRebalance` custom resource with the `spec.mode` field as `add-broker` and the list of the new brokers in the `spec.brokers` field.
+3. The `KafkaRebalanceAssemblyOperator` starts the interaction with Cruise Control via the `/add_broker` endpoint for getting an optimization proposal (by using the dryrun feature).
+4. The user accepts the proposal by applying the `strimzi.io/rebalance=approve` annotation on it.
+5. The `KafkaRebalanceAssemblyOperator` starts the interaction with Cruise Control via the `/add_broker` endpoint for running the actual rebalancing.
+
+#### Rebalance on scale down
+
+When scaling down the cluster, the user can go through the following procedure:
+
+1. The user creates a `KafkaRebalance` custom resource with the `spec.mode` field as `remove-broker` and the list of the target brokers to remove in the `spec.brokers` field.
```

Comment:
Again, as with the scale up, I think Strimzi should create (if CC is configured) a KRCR with the default settings and the brokers field filled in. With StatefulSets we know which brokers are to be removed. I am not sure how the PodSet scale down works but presumable the user has to say which brokers to remove (or strimzi decides) so we should be able to create a KRCR for them.

Reply:
We can block the scale down until the remove broker KRCR is in `Ready` state.